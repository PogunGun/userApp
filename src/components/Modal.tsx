import {Box, Button,Modal,TextField} from "@mui/material";
import { useState} from "react";
import ReactDom from "react-dom";
import {Controller, SubmitHandler, useForm, useFormState} from "react-hook-form";
import {nameValidation, usernameValidation} from "./validation";

interface ISignInForm {
    name: string;
    username: string;
}


function ModalBlock({open,setUserTitle, setTitle, updateUser,setOpen,id}: any) {
    const {control} = useForm<ISignInForm>({
        mode: 'onChange',
    });

    const {errors} = useFormState({
        control
    })

    const handleClose = (e:any) =>{


        setOpen(false)
    }

    const portalDiv = document.getElementById('portal')!;

    return ReactDom.createPortal(
        <Modal
            onClick={event =>event.stopPropagation()}
            open={open}
            onClose={(e:any)=>handleClose(e)}
            aria-labelledby="modal-modal-title"
            aria-describedby="modal-modal-description"
        >
            <Box sx={style.box}>
                <Box
                    component="form"
                    sx={{
                        '& > :not(style)': {m: 1},
                    }}
                    autoComplete="off"
                >
                    <Controller
                        control={control}
                        name="name"
                        rules={nameValidation}
                        render={({field}) => (
                            <TextField
                                label="Name"
                                onChange={(e) => {
                                    field.onChange(e)
                                    setTitle(e.target.value)
                                }
                                }
                                value={field.value}
                                fullWidth={true}
                                size="small"
                                margin="normal"
                                className="auth-form__input"
                                error={!!errors.name?.message}
                                helperText={errors?.name?.message}
                            />
                        )}
                    />
                    <Controller
                        control={control}
                        name="username"
                        rules={usernameValidation}
                        render={({field}) => (
                            <TextField
                                label="UserName"
                                onChange={(e) => {
                                    field.onChange(e)
                                    setUserTitle(e.target.value)
                                }}
                                fullWidth={true}
                                size="small"
                                margin="normal"
                                className="auth-form__input"
                                error={!!errors?.username?.message}
                                helperText={errors?.username?.message}
                            />
                        )}
                    />
                    <Button onClick={(e) =>{
                        updateUser(id)}
                    }>Change User</Button>
                </Box>
            </Box>
        </Modal>,portalDiv
)
}

export const style = {
    box: {
        position: 'absolute' as 'absolute',
        top: '50%',
        left: '50%',
        transform: 'translate(-50%, -50%)',
        width: 400,
        bgcolor: 'background.paper',
        boxShadow: 24,
        p: 4,
    }
}
export default ModalBlock;

