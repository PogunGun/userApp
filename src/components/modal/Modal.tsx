import {
  Box, Button, Modal, TextField,
} from '@mui/material';
import React, { FC, useContext } from 'react';
import ReactDom from 'react-dom';
import {
  Controller, useForm, useFormState,
} from 'react-hook-form';
import { nameValidation, usernameValidation } from '../validation/validation';
import { ColorButton } from '../material/searchAppBar';
import { ISignInForm, IModal } from '../../type/types';

import { MyContext } from '../MainScreen/MainScreen';

export const style = {
  box: {
    position: 'absolute' as 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)',
    width: 400,
    bgcolor: 'background.paper',
    boxShadow: 24,
    p: 4,
  },
};

export const ModalBlock: FC<IModal> = ({
  open, updateUser, setOpen, id, create,
}) => {
  const { control } = useForm<ISignInForm>({
    mode: 'onChange',
  });
  const {
    setTitle, setUserTitle, title, userTitle,
  } = useContext(MyContext);
  const { errors } = useFormState({
    control,
  });

  const handleClose = () => {
    setOpen(false);
  };

  const portalDiv = document.getElementById('portal')!;

  return ReactDom.createPortal(<Modal
            onClick={(event: React.MouseEvent) => event.stopPropagation()}
            open={open}
            onClose={() => handleClose()}
            aria-labelledby="modal-modal-title"
            aria-describedby="modal-modal-description"
        >
            <Box sx={style.box}>
                <Box
                    component="form"
                    sx={{
                      '& > :not(style)': { m: 1 },
                    }}
                    autoComplete="off"
                >
                    <Controller
                        control={control}
                        name="name"
                        rules={nameValidation}
                        render={({ field }) => (
                            <TextField
                                label="Name"
                                onChange={(e) => {
                                  field.onChange(e);
                                  setTitle(e.target.value);
                                }
                                }
                                value={title}
                                fullWidth={true}
                                size="small"
                                margin="normal"
                                className="auth-form__input"
                                error={!!errors.name?.message}
                                helperText={errors?.name?.message}
                            />
                        )}
                    />
                    <Controller
                        control={control}
                        name="username"
                        rules={usernameValidation}
                        render={({ field }) => (
                            <TextField
                                label="UserName"
                                onChange={(e) => {
                                  field.onChange(e);
                                  setUserTitle(e.target.value);
                                }}
                                value={userTitle}
                                fullWidth={true}
                                size="small"
                                margin="normal"
                                className="auth-form__input"
                                error={!!errors?.username?.message}
                                helperText={errors?.username?.message}
                            />
                        )}
                    />

                    {(create != null)
                      ? <ColorButton onClick={() => create()}>Create User</ColorButton>
                      : <Button onClick={() => {
                        if (updateUser != null) {
                          if (typeof id === 'number') {
                            updateUser(id);
                          }
                        }
                      }}>Change User</Button>

                    }
                </Box>
            </Box>
        </Modal>, portalDiv);
};
