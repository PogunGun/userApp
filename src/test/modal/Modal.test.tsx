import { render, screen } from '@testing-library/react';
import React from 'react';
import { ModalBlock } from '../../components/modal/Modal';

const handelClick = jest.fn();

describe('Modal Test', () => {
  test('INPUT CREATE', async () => {
    const { asFragment } = render(
            <ModalBlock open={true} setOpen={() => false}/>,
    );
    const inputName = await screen.findByTestId('name-input');
    const inputUserName = await screen.findByTestId('username-input');
    const update = await screen.findByTestId('update');
    expect(update)
      .toBeInTheDocument();
    expect(inputName)
      .toBeInTheDocument();
    expect(inputUserName)
      .toBeInTheDocument();
    expect(asFragment())
      .toMatchSnapshot();
  });
  test('Update Click', async () => {
    render(
            <ModalBlock open={true} setOpen={() => false} updateUser={handelClick} id={1}/>,
    );
    const update = await screen.findByTestId('update');
    expect(update)
      .toBeInTheDocument();
    update.click();
    expect(handelClick)
      .toBeCalled();
  });
  test('Create click', async () => {
    render(
            <ModalBlock open={true} setOpen={() => false} create={handelClick}/>,
    );
    const create = await screen.findByTestId('create');
    expect(create)
      .toBeInTheDocument();
    create.click();
    expect(handelClick)
      .toBeCalled();
  });
});
