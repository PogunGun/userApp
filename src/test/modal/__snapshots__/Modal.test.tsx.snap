// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainScreen Test Create click 1`] = `<DocumentFragment />`;

exports[`MainScreen Test INPUT CREATE 1`] = `<DocumentFragment />`;

exports[`MainScreen Test Update Click 1`] = `<DocumentFragment />`;
